### PR TITLE
Add partial circular import fix (and test) for #4390

### PR DESCRIPTION
**Pull request by SyamGadde**
Proposed fix for #4390.

### CODE REVIEW

File: tests/run/partial_circular_import.srctree [hunk 1]
```diff
@@ -0,0 +1,23 @@
+PYTHON setup.py build_ext --inplace
```
Comment (da-woods) [inline tests/run/partial_circular_import.srctree:1]:
```suggestion
PYTHON -c "import pkg.A"
PYTHON setup.py build_ext --inplace
```

For these tests where we're aiming to match the behaviour to Python we often try to run the test in Python too.

File: Cython/Utility/ImportExport.c [hunk 2]
```diff
@@ -238,6 +238,34 @@ static PyObject* __Pyx_ImportFrom(PyObject* module, PyObject* name); /*proto*/
...
+        PyObject* module_dot = 0;
+        PyObject* full_name = 0;
+        PyErr_Clear();
+        module_name_str = PyModule_GetName(module);
+        if (unlikely(!module_name_str)) { goto modbad; }
+        module_name = PyUnicode_FromString(module_name_str);
+        if (unlikely(!module_name)) { goto modbad; }
+        dot = PyUnicode_FromString(".");
+        if (unlikely(!dot)) { goto modbad; }
+        module_dot = PyUnicode_Concat(module_name, dot);
```
Comment (da-woods) [inline Cython/Utility/ImportExport.c:256]:
```suggestion
        module_dot = PyUnicode_Concat(module_name, PYUNICODE("."));
```

This is a shortcut Cython supports for utility code for Unicode constants - it gets substituted when the source file is generated and just avoids you having to manage them manually. You then don't need to manage `PyObject *dot` (so can remove some of your lines above)

**Review by da-woods — COMMENTED** [review]
Thanks for creating this. I've made a couple of small suggestions

**Comment by da-woods** [comment]
I'd also add `partial_circular_import` to pypy2_bugs.txt - I suspect there isn't an easy way to get it to work.

File: Cython/Utility/ImportExport.c (hunk 2 above)
Comment (SyamGadde) [reply Cython/Utility/ImportExport.c]: Thanks, committed changes!

File: tests/run/partial_circular_import.srctree [hunk 3]
```diff
@@ -0,0 +1,24 @@
+PYTHON -c "import pkg.A"
```
Comment (da-woods) [inline tests/run/partial_circular_import.srctree:1]: It looks like this requires a somewhat recent version of Python for the Python tests to work. (It's not a concern for the Cython side of it)

```suggestion
PYTHON -c "import sys; if sys.version_info >= (3,5): import pkg.A"
```

Comment (SyamGadde) [reply tests/run/partial_circular_import.srctree]: Looks like the presence of pkg/__init__.py causes the problem for some reason (but it didn't seem to be a problem in the first round of commits).

**Comment by da-woods** [comment]
@scoder this might be a good candidate to include in the next 3.0 alpha release. I believe this is a regression from 0.29.x

File: tests/run/partial_circular_import.srctree [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+######## pkg/__init__.py ########
+
+######## pkg/A.py ########
+from . import B
+
+######## pkg/B.py ########
+from . import C
+
+######## pkg/C.py ########
+from . import B
```
Comment (scoder) [inline tests/run/partial_circular_import.srctree:23]:
```suggestion
from . import B

def verify(rel_B):
    import pkg.B
    assert pkg.B == rel_B

verify(B)
```

File: tests/run/partial_circular_import.srctree [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    ext_modules = cythonize("**/*.py"),
+    )
+
+######## pkg/__init__.py ########
+
+######## pkg/A.py ########
+from . import B
+
+######## pkg/B.py ########
+from . import C
```
Comment (scoder) [inline tests/run/partial_circular_import.srctree:20]:
```suggestion
from . import C

def verify(rel_C):
    import pkg.C
    assert pkg.C == rel_C
verify(C)
```

File: tests/run/partial_circular_import.srctree [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+from distutils.core import setup
+
+setup(
+    ext_modules = cythonize("**/*.py"),
+    )
+
+######## pkg/__init__.py ########
+
+######## pkg/A.py ########
+from . import B
```
Comment (scoder) [inline tests/run/partial_circular_import.srctree:17]:
```suggestion
from . import B

def verify(rel_B):
    import pkg.B
    assert pkg.B == rel_B
verify(B)
```

File: tests/run/partial_circular_import.srctree [hunk 7]
```diff
@@ -16,9 +16,25 @@ setup(
...
+verify(C)
+
 ######## pkg/C.py ########
 from . import B
 
+def verify(rel_B):
+    import pkg.B
+    assert pkg.B == rel_B
+
+verify(B)
```
Comment (da-woods) [inline tests/run/partial_circular_import.srctree:39]: I don't think this test can work (hence it fails on pure Python too) - the contents of `B` aren't filled in until after module `C` has run fully. I think maybe you should call verify up in the `PYTHON -c ...` lines after the import statement has completed fully.

Comment (scoder) [reply tests/run/partial_circular_import.srctree:39]: Interesting. I would have expected that to work…

Anyway, what I wanted to assure is that `__Pyx_Import()` returns the module and not the package.

I'm not even sure that we need `__Pyx_Import()` here. A lookup in `sys.modules` might be enough. Didn't take the time to dig up what exactly Python is doing here – in any case, we should mimic that.

Comment (da-woods) [reply tests/run/partial_circular_import.srctree:39]: Having tested it in more detail: I think I misrepresented the problem with these tests slightly. It's actually that the contents of `pkg` aren't filled in until all its submodules are successfully reported

As you suspect, Python is just looking up in `sys.modules` - https://github.com/python/cpython/commit/0373a106a1ca9ce67cd99bf0a93ab7c3abd15cb6 (that's the implementation from 3.5 but the current version isn't notably different) so it's probably worth simplifying to that

File: tests/run/partial_circular_import.srctree [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+PYTHON setup.py build_ext --inplace
+PYTHON -c "import pkg.A"
+
+######## setup.py ########
+
+from Cython.Build.Dependencies import cythonize
+from distutils.core import setup
+
+setup(
+    ext_modules = cythonize("**/*.py"),
```
Comment (scoder) [inline tests/run/partial_circular_import.srctree:11]:
```suggestion
    ext_modules = cythonize("*/*.py"),
```

File: tests/run/partial_circular_import.srctree [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    )
+
+######## pkg/__init__.py ########
+
+######## pkg/A.py ########
+from . import B
+
+def verify(rel_B):
+    import pkg.B
+    assert pkg.B == rel_B
```
Comment (scoder) [inline tests/run/partial_circular_import.srctree:21]:
```suggestion
    import pkg.B as abs_B
    assert abs_B == rel_B
```

File: tests/run/partial_circular_import.srctree [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+verify(B)
+
+######## pkg/B.py ########
+from . import C
+
+def verify(rel_C):
+    import pkg.C
+    assert pkg.C == rel_C
```
Comment (scoder) [inline tests/run/partial_circular_import.srctree:29]:
```suggestion
    import pkg.C as abs_C
    assert abs_C == rel_C
```

File: tests/run/partial_circular_import.srctree [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+verify(C)
+
+######## pkg/C.py ########
+from . import B
+
+def verify(rel_B):
+    import pkg.B
+    assert pkg.B == rel_B
```
Comment (scoder) [inline tests/run/partial_circular_import.srctree:37]:
```suggestion
    import pkg.B as abs_B
    assert abs_B == rel_B
```

File: Cython/Utility/ImportExport.c [hunk 12]
```diff
@@ -238,6 +238,40 @@ static PyObject* __Pyx_ImportFrom(PyObject* module, PyObject* name); /*proto*/
...
+        if (unlikely(!module_dot)) { goto modbad; }
+        full_name = PyUnicode_Concat(module_dot, name);
+        if (unlikely(!full_name)) { goto modbad; }
+        #if PY_VERSION_HEX < 0x030700A1 || (CYTHON_COMPILING_IN_PYPY && PYPY_VERSION_NUM  < 0x07030400)
+        {
+            PyObject *modules = PyImport_GetModuleDict();
+            if (unlikely(!modules))
+                goto modbad;
+            value = __Pyx_PyDict_GetItemStr(modules, full_name);
+            Py_XINCREF(value);
```
Comment (da-woods) [inline Cython/Utility/ImportExport.c:263]:
```suggestion
            value = PyObject_GetItem(modules, full_name);
```

I genuinely don't understand why, but this seems to make it pass on Python 3.6 (for me at least)

Comment (scoder) [reply Cython/Utility/ImportExport.c:263]: The reason is that `__Pyx_PyDict_GetItemStr()` assumes a known string hash value (usually for a literal string). Here, we are creating a new string for the FQMN before we call it. That's a bug.

File: tests/run/partial_circular_import.srctree [hunk 13]
```diff
@@ -0,0 +1,40 @@
+PYTHON -c 'if __import__("sys").version_info >= (3,7): import pkg.A'
```
Comment (scoder) [inline tests/run/partial_circular_import.srctree:1]:
```suggestion
PYTHON -c 'if __import__("sys").version_info >= (3,5): import pkg.A'
```

**Pull request merged**